packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3070,12 +3070,33 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
       return null;
     }
     if (child != null) {
-      if (child.widget == newWidget) {
+      bool canUpdate = true;
+      // When the type of a widget is changed between Stateful and Stateless via
+      // hot reload, the element tree will end up in a partially invalid state.
+      // That is, if the widget was a StatefulWidget and is now a StatelessWidget,
+      // then the element tree currently contains a StatefulElement that is incorrectly
+      // referencing a StatelessWidget (and likewise with StatelessElement).
+      //
+      // To avoid crashing due to type errors, we need to gently guide the invalid
+      // element out of the tree. To do so, we ensure that the `canUpdate` condition
+      // returns false which prevents us from trying to update the existing element
+      // incorrectly.
+      //
+      // For the case where the widget becomes Stateful, we also need to avoid
+      // accessing `StatelessElement.widget` as the cast on the getter will
+      // cause a type error to be thrown. Here we avoid that by short-circuiting
+      // the `Widget.canUpdate` check once `canUpdate` is false.
+      assert(() {
+        canUpdate = !((child is StatelessElement && newWidget is StatefulWidget) ||
+                      (child is StatefulElement && newWidget is StatelessWidget));
```

Comment:
since all the interesting logic is in debug-mode only code, you could just go crazy and do something like:

```dart
int oldWidgetClass = child is StatelessElement ? 1 :
                     child is StatefulElement ? 2 :
                     child is SingleChildRenderObjectElement ? 3 :
                     child is InheritedElement ? 4 :
                     0;
int newWidgetClass = newWidget is StatelessElement ? 1 :
                     newWidget is StatefulElement ? 2 :
                     newWidget is SingleChildRenderObjectElement ? 3 :
                     newWidget is InheritedElement ? 4 :
                     0;
hasSameSuperlass = oldWidgetClass == newWidgetClass;
```

(but with all the relevant subclasses. And maybe put that common superclass-identifying logic in a private method on Widget or something.)

Reply:
I refactored both into static helpers on the Widget and Element subclasses. It would be nice if we had a generic sort of mapping between an Element subtype and the particular Widget subtype it contains, but that doesn't seem possible based on the fairly open-ended nature of the APIs today.

Given that we've only ever had this reported as an error for stateless/stateful, I've only included those and the suggestion addition of SingleChild/Inherited